pandas/core/case_when.py

```diff
@@ -0,0 +1,212 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+import numpy as np
+
+from pandas._libs import lib
+
+from pandas.core.dtypes.cast import (
+    construct_1d_arraylike_from_scalar,
+    find_common_type,
+    infer_dtype_from,
+)
+from pandas.core.dtypes.common import is_scalar
+from pandas.core.dtypes.generic import ABCSeries
+
+from pandas.core.construction import array as pd_array
+
+if TYPE_CHECKING:
+    from pandas._typing import (
+        ArrayLike,
+        Scalar,
+        Series,
+    )
+
+
+def case_when(
+    *args: tuple[tuple[tuple[ArrayLike, ArrayLike | Scalar]]],
```

Comment:
I still think this is incorrect. I think the correct type here is `tuple[ArrayLike, ArrayLike | Scalar]`.

Here's a simple example that illustrates how to type a `*args` argument, where each element is a tuple:
```python
def foobar(*args: tuple[int, str]):
    print(args)


foobar((3, "abc"), (5, "def"))
```
The above type checks correctly.

Having said that, I think a better design that corresponds to how we do other pandas operations would be to pass a list of these tuples.  In your example, you have:
```python
    pd.case_when(
        (df.a == 1, 'first'),  # condition, replacement
        (df.a.gt(1) & df.b.eq(5), 'second'),  # condition, replacement
        default = 'default', # optional
   )
```
I think a better design would be to have users put the cases in a list, e.g.:
```python
    pd.case_when([
        (df.a == 1, 'first'),  # condition, replacement
        (df.a.gt(1) & df.b.eq(5), 'second')],  # condition, replacement
        default = 'default', # optional
   )
```
This is just a matter of personal taste - other pandas maintainers may prefer your design.  I'm just not a big fan of using `*args` and I don't think we have any public method that has `*args` as the first argument.   So then your function declaration would be:
```python
def case_when(
     caselist: list[tuple[ArrayLike, ArrayLike | Scalar]],
     default: ArrayLike | Scalar = lib.no_default,
) -> Series:
```
Then it becomes clearer, from the argument name, that you are passing a list of cases.